rfcs/2020-04-04-2221-kubernetes-integration.md

```diff
@@ -0,0 +1,1625 @@
+- Suggest setting up proper RBAC rules for cluster operators and users;
+  [`audit2rbac`](https://github.com/liggitt/audit2rbac) is a useful tool to
+  help with this.
+- Suggest using [Pod Security Policies][k8s_docs_pod_security_policiy]
+  ([API][k8s_api_pod_security_policy]).
+- Suggest using [NetworkPolicy][k8s_api_network_policy].
+- Suggest runnig [kube-bench].
+- Suggest reading the
+  [Kubernetes security documentation][k8s_docs_securing_a_cluster].
+
+#### Automatic container rebuilds
+
+The ability to rebuild containers with a CVE fix automatically quickly is a very
+important part of a successful vulnerability mitigation strategy.
+We should prepare in advance and rollout the infrastructure and automation to
+make it possible to rebuild the containers for _all_ (not just the latest or
+nightly!) the supported Vector versions.
+
+## Prior Art
+
+1. [Filebeat k8s integration]
+1. [Fluentbit k8s integration]
+1. [Fluentd k8s integration]
+1. [LogDNA k8s integration]
+1. [Honeycomb integration]
+1. [Bonzai logging operator] - This is approach is likely outside of the scope
+   of Vector's initial Kubernetes integration because it focuses more on
+   deployment strategies and topologies. There are likely some very useful
+   and interesting tactics in their approach though.
+1. [Influx Helm charts]
+1. [Awesome Operators List] - an "awesome list" of operators.
+
+## Sales Pitch
+
+See [motivation](#motivation).
+
+## Drawbacks
+
+1. Increases the surface area that our team must manage.
+
+## Alternatives
+
+1. Not do this integration and rely solely on external community-driven
+   integrations.
+
+## Outstanding Questions
+
+### From Ben
+
+1. ~~What is the minimal Kubernetes version that we want to support. See
+   [this comment][kubernetes_version_comment].~~
+   See the [Minimal supported Kubernetes version][anchor_minimal_supported_kubernetes_version]
+   section.
+1. ~~What is the best to avoid Vector from ingesting it's own logs? I'm assuming
+   that my [`kubectl` tutorial][anchor_tutorial_kubectl] handles this with
+   namespaces?
+   We'd just need to configure Vector to exclude this namespace?~~
+   See the [Origin filtering][anchor_origin_filtering] section.
+1. ~~I've seen two different installation strategies. For example, Fluentd offers
+   a [single daemonset configuration file][fluentd_daemonset] while Fluentbit
+   offers [four separate configuration files][fluentbit_installation]
+   (`service-account.yaml`, `role.yaml`, `role-binding.yaml`, `configmap.yaml`).
+   Which approach is better? Why are they different?~~
+   See the
+   [Strategy on YAML file grouping][anchor_strategy_on_yaml_file_grouping]
+   section.
+1. ~~Should we prefer `kubectl create ...` or `kubectl apply ...`? The examples
+   in the [prior art](#prior-art) section use both.~~
+   See [Helm vs raw YAML files][anchor_helm_vs_raw_yaml_files] section.
+1. ~~From what I understand, Vector requires the Kubernetes `watch` verb in order
```

Comment:
Did we address this? Is there anything special we need to include in the docs regarding k8s permissions? Or is that handled with the YAML files automatically? If there is, we should add this requirement in the plan of attack.

Reply:
I view this as a security concern, see https://github.com/timberio/vector/blob/k8s-rfc/rfcs/2020-04-04-2221-kubernetes-integration.md#deployment-hardening

In the plan of attack, it's part of both preparing YAML files / Heml Charts and altering `kubernetes` source. It won't work without proper config, so don't worry - we won't miss it.

Reply:
Cool. I was mostly thinking about the requirement for [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/rbac/), which means that the Kubernetes API must be started with the appropriate flags:

```
kube-apiserver --authorization-mode=Example,RBAC --other-options --more-options
````

I assume that's something we'll need to document. I'm sure you've got this covered though 😄 .

Reply:
This is sorta the default for a long time now, but yeah, we should probably leave a note about RBAC.

Reply:
Added a note on RBAC to the RFC.